File: src/code/goal_stack_planning.java
import java.util.*;

// -------------------- Predicate Class --------------------
class Predicate {
    String name, X, Y;

    Predicate(String name, String x, String y) {
        this.name = name;
        this.X = x;
        this.Y = y;
    }

    Predicate(String name) {
        this(name, "", "");
    }

    @Override
    public String toString() {
        if (X.equals("") && Y.equals("")) return name;
        if (Y.equals("")) return name + "(" + X + ")";
        return name + "(" + X + "," + Y + ")";
    }

    @Override
    public boolean equals(Object o) {
        if (!(o instanceof Predicate)) return false;
        return this.toString().equals(o.toString());
    }

    @Override
    public int hashCode() {
        return toString().hashCode();
    }
}

// ------------------ Action Class -------------------
class Action {
    String name, X, Y;
    Set<Predicate> preconds = new HashSet<>();
    Set<Predicate> add = new HashSet<>();
    Set<Predicate> del = new HashSet<>();

    Action(String name, String x, String y) {
        this.name = name;
        this.X = x;
        this.Y = y;
    }

    Action(String name, String x) {
        this(name, x, "");
    }

    @Override
    public String toString() {
        if (X.equals("") && Y.equals("")) return name;
        if (Y.equals("")) return name + "(" + X + ")";
        return name + "(" + X + "," + Y + ")";
    }
}

// -------------------- Helpers for Predicates --------------------
class Predicates {
    static Predicate ON(String x, String y) { return new Predicate("ON", x, y); }
    static Predicate ONTABLE(String x) { return new Predicate("ONTABLE", x, ""); }
    static Predicate CLEAR(String x) { return new Predicate("CLEAR", x, ""); }
    static Predicate HOLDING(String x) { return new Predicate("HOLDING", x, ""); }
    static Predicate ARMEMPTY() { return new Predicate("ARMEMPTY", "", ""); }
}

// -------------------- Action Generators --------------------
class Actions {
    static Action PICKUP(String x) {
        Action a = new Action("PICKUP", x);
        a.preconds.add(Predicates.ONTABLE(x));
        a.preconds.add(Predicates.CLEAR(x));
        a.preconds.add(Predicates.ARMEMPTY());
        a.add.add(Predicates.HOLDING(x));
        a.del.add(Predicates.ONTABLE(x));
        a.del.add(Predicates.CLEAR(x));
        a.del.add(Predicates.ARMEMPTY());
        return a;
    }

    static Action PUTDOWN(String x) {
        Action a = new Action("PUTDOWN", x);
        a.preconds.add(Predicates.HOLDING(x));
        a.add.add(Predicates.ONTABLE(x));
        a.add.add(Predicates.CLEAR(x));
        a.add.add(Predicates.ARMEMPTY());
        a.del.add(Predicates.HOLDING(x));
        return a;
    }

    static Action UNSTACK(String x, String y) {
        Action a = new Action("UNSTACK", x, y);
        a.preconds.add(Predicates.ON(x, y));
        a.preconds.add(Predicates.CLEAR(x));
        a.preconds.add(Predicates.ARMEMPTY());
        a.add.add(Predicates.HOLDING(x));
        a.add.add(Predicates.CLEAR(y));
        a.del.add(Predicates.ON(x, y));
        a.del.add(Predicates.CLEAR(x));
        a.del.add(Predicates.ARMEMPTY());
        return a;
    }

    static Action STACK(String x, String y) {
        Action a = new Action("STACK", x, y);
        a.preconds.add(Predicates.HOLDING(x));
        a.preconds.add(Predicates.CLEAR(y));
        a.add.add(Predicates.ON(x, y));
        a.add.add(Predicates.CLEAR(x));
        a.add.add(Predicates.ARMEMPTY());
        a.del.add(Predicates.HOLDING(x));
        a.del.add(Predicates.CLEAR(y));
        return a;
    }
}

// -------------------- GSP Algorithm --------------------
public class GoalStackPlanner {

    static Action chooseAction(Predicate g, Set<Predicate> state, List<String> blocks) {
        if (g.name.equals("ON")) return Actions.STACK(g.X, g.Y);
        if (g.name.equals("ONTABLE")) return Actions.PUTDOWN(g.X);
        if (g.name.equals("CLEAR")) {
            for (String b : blocks) {
                if (state.contains(Predicates.ON(b, g.X))) 
                    return Actions.UNSTACK(b, g.X);
            }
            return Actions.PUTDOWN(g.X);
        }
        if (g.name.equals("HOLDING")) {
            if (state.contains(Predicates.ONTABLE(g.X))) return Actions.PICKUP(g.X);
            for (String b : blocks) {
                if (state.contains(Predicates.ON(g.X, b)))
                    return Actions.UNSTACK(g.X, b);
            }
            return Actions.PICKUP(g.X);
        }
        if (g.name.equals("ARMEMPTY")) {
            for (String b : blocks) {
                if (state.contains(Predicates.HOLDING(b)))
                    return Actions.PUTDOWN(b);
            }
            return Actions.PUTDOWN(blocks.get(0));
        }
        return null;
    }

    public static List<Action> plan(Set<Predicate> init, Set<Predicate> goal, List<String> blocks) {
        Set<Predicate> state = new HashSet<>(init);
        Stack<List<Predicate>> stack = new Stack<>();
        stack.push(new ArrayList<>(goal));
        List<Action> plan = new ArrayList<>();

        System.out.println("\n===== INITIAL STATE =====");
        state.forEach(s -> System.out.println(s));

        System.out.println("\n===== GOAL STATE =====");
        goal.forEach(g -> System.out.println(g));

        System.out.println("\n===== PLANNING PROCESS =====");

        while (!stack.isEmpty()) {
            List<Predicate> top = stack.pop();

            if (top.size() > 1) { // Conjunction
                List<Predicate> unsatisfied = new ArrayList<>();
                for (Predicate g : top)
                    if (!state.contains(g)) unsatisfied.add(g);

                if (!unsatisfied.isEmpty()) {
                    stack.push(top);
                    for (Predicate g : unsatisfied)
                        stack.push(Collections.singletonList(g));

                    System.out.print("→ Unsatisfied goals: ");
                    unsatisfied.forEach(x -> System.out.print(x + " "));
                    System.out.println();
                }
            } else { // Single Goal
                Predicate g = top.get(0);
                if (state.contains(g)) {
                    System.out.println("✓ Already satisfied: " + g);
                    continue;
                }

                Action action = chooseAction(g, state, blocks);
                System.out.println("→ Choosing action: " + action);

                boolean precondsMet = true;
                for (Predicate p : action.preconds)
                    if (!state.contains(p)) precondsMet = false;

                if (!precondsMet) {
                    System.out.print("→ Pushing missing preconditions: ");
                    stack.push(top);
                    for (Predicate p : action.preconds) {
                        if (!state.contains(p)) {
                            stack.push(Collections.singletonList(p));
                            System.out.print(p + " ");
                        }
                    }
                    System.out.println();
                } else {
                    state.removeAll(action.del);
                    state.addAll(action.add);
                    plan.add(action);

                    System.out.println("✓ Executing: " + action);
                    System.out.print("Updated State: { ");
                    state.forEach(s -> System.out.print(s + " "));
                    System.out.println("}");
                }
            }
        }
        return plan;
    }

    // -------------------- MAIN --------------------
    public static void main(String[] args) {
        List<String> blocks = Arrays.asList("A", "B", "C");

        Set<Predicate> initial = new HashSet<>(Arrays.asList(
            Predicates.ONTABLE("A"),
            Predicates.ON("B", "A"),
            Predicates.ONTABLE("C"),
            Predicates.CLEAR("B"),
            Predicates.CLEAR("C"),
            Predicates.ARMEMPTY()
        ));

        Set<Predicate> goal = new HashSet<>(Arrays.asList(
            Predicates.ON("A", "B"),
            Predicates.ON("B", "C"),
            Predicates.ONTABLE("C"),
            Predicates.ARMEMPTY()
        ));

        List<Action> finalPlan = plan(initial, goal, blocks);

        System.out.println("\n===== FINAL PLAN =====");
        int step = 1;
        for (Action a : finalPlan)
            System.out.println(step++ + ". " + a);
    }
}

File: src/code/goal_stack_planning.cpp
#include <iostream>
#include <vector>
#include <stack>
#include <set>
#include <string>
#include <algorithm>
using namespace std;

// -------------------- Predicate Structure --------------------
struct Predicate {
    string name, X, Y;

    Predicate(string n, string x = "", string y = "") : name(n), X(x), Y(y) {}

    string str() const {
        if (X == "" && Y == "") return name;
        if (Y == "") return name + "(" + X + ")";
        return name + "(" + X + "," + Y + ")";
    }

    bool operator<(const Predicate& other) const {
        return str() < other.str();
    }
};

// -------------------- Helper Functions for Predicates --------------------
Predicate ON(string x, string y) { return Predicate("ON", x, y); }
Predicate ONTABLE(string x) { return Predicate("ONTABLE", x); }
Predicate CLEAR(string x) { return Predicate("CLEAR", x); }
Predicate HOLDING(string x) { return Predicate("HOLDING", x); }
Predicate ARMEMPTY() { return Predicate("ARMEMPTY"); }

// -------------------- Action Structure --------------------
struct Action {
    string name, X, Y;
    set<Predicate> preconds, add, del;

    Action(string n, string x = "", string y = "") : name(n), X(x), Y(y) {}

    string str() const {
        if (X == "" && Y == "") return name;
        if (Y == "") return name + "(" + X + ")";
        return name + "(" + X + "," + Y + ")";
    }
};

// -------------------- Action Generators --------------------
Action PICKUP(string x) {
    Action a("PICKUP", x);
    a.preconds = { ONTABLE(x), CLEAR(x), ARMEMPTY() };
    a.add = { HOLDING(x) };
    a.del = { ONTABLE(x), CLEAR(x), ARMEMPTY() };
    return a;
}

Action PUTDOWN(string x) {
    Action a("PUTDOWN", x);
    a.preconds = { HOLDING(x) };
    a.add = { ONTABLE(x), CLEAR(x), ARMEMPTY() };
    a.del = { HOLDING(x) };
    return a;
}

Action UNSTACK(string x, string y) {
    Action a("UNSTACK", x, y);
    a.preconds = { ON(x, y), CLEAR(x), ARMEMPTY() };
    a.add = { HOLDING(x), CLEAR(y) };
    a.del = { ON(x, y), CLEAR(x), ARMEMPTY() };
    return a;
}

Action STACK(string x, string y) {
    Action a("STACK", x, y);
    a.preconds = { HOLDING(x), CLEAR(y) };
    a.add = { ON(x, y), CLEAR(x), ARMEMPTY() };
    a.del = { HOLDING(x), CLEAR(y) };
    return a;
}

// -------------------- Goal Stack Planner --------------------
Action choose_action(Predicate g, set<Predicate>& state, vector<string>& blocks) {
    if (g.name == "ON") return STACK(g.X, g.Y);
    if (g.name == "ONTABLE") return PUTDOWN(g.X);
    if (g.name == "CLEAR") {
        for (string b : blocks)
            if (state.count(ON(b, g.X))) return UNSTACK(b, g.X);
        return PUTDOWN(g.X);
    }
    if (g.name == "HOLDING") {
        if (state.count(ONTABLE(g.X))) return PICKUP(g.X);
        for (string b : blocks)
            if (state.count(ON(g.X, b))) return UNSTACK(g.X, b);
        return PICKUP(g.X);
    }
    if (g.name == "ARMEMPTY") {
        for (string b : blocks)
            if (state.count(HOLDING(b))) return PUTDOWN(b);
        return PUTDOWN(blocks[0]);  // fallback
    }
    return Action("NO_ACTION");
}

// -------------------- Main Planner --------------------
vector<Action> goal_stack_planner(set<Predicate> initial_state,
                                  set<Predicate> goal_state,
                                  vector<string>& blocks) {

    set<Predicate> state = initial_state;
    stack<vector<Predicate>> st;
    vector<Predicate> goal_list(goal_state.begin(), goal_state.end());
    st.push(goal_list);
    vector<Action> plan;

    cout << "\n========= INITIAL STATE =========\n";
    for (auto& s : state) cout << s.str() << endl;

    cout << "\n========= GOAL STATE =========\n";
    for (auto& g : goal_state) cout << g.str() << endl;

    cout << "\n========= PLANNING PROCESS =========\n";

    while (!st.empty()) {
        auto top = st.top(); st.pop();

        if (top.size() > 1) { // Conjunction
            vector<Predicate> unsatisfied;
            for (auto& p : top)
                if (!state.count(p)) unsatisfied.push_back(p);

            if (!unsatisfied.empty()) {
                st.push(top);
                for (auto& p : unsatisfied) st.push({ p });
                cout << "→ Unsatisfied goals:";
                for (auto& x : unsatisfied) cout << " " << x.str();
                cout << endl;
            }
        }
        else { // Single goal
            Predicate g = top[0];
            if (state.count(g)) {
                cout << "✓ Already satisfied: " << g.str() << endl;
                continue;
            }

            Action act = choose_action(g, state, blocks);
            cout << "→ Choosing action: " << act.str() << endl;

            // Check preconditions
            bool ok = true;
            for (auto& p : act.preconds)
                if (!state.count(p)) ok = false;

            if (!ok) {  // Push preconditions
                cout << "→ Preconditions not met, pushing:";
                st.push({ g });
                for (auto& p : act.preconds) {
                    if (!state.count(p)) {
                        st.push({ p });
                        cout << " " << p.str();
                    }
                }
                cout << endl;
            }
            else {
                // Apply action
                for (auto& d : act.del) state.erase(d);
                for (auto& a : act.add) state.insert(a);
                plan.push_back(act);

                cout << "✓ Executing: " << act.str() << endl;
                cout << "Updated State: { ";
                for (auto& s : state) cout << s.str() << " ";
                cout << "}\n";
            }
        }
    }
    return plan;
}

// -------------------- MAIN FUNCTION --------------------
int main() {
    vector<string> blocks = {"A", "B", "C"};

    // INITIAL STATE
    set<Predicate> initial_state = {
        ONTABLE("A"), ON("B","A"), ONTABLE("C"),
        CLEAR("B"), CLEAR("C"), ARMEMPTY()
    };

    // GOAL STATE
    set<Predicate> goal_state = {
        ON("A","B"), ON("B","C"), ONTABLE("C"), ARMEMPTY()
    };

    vector<Action> plan = goal_stack_planner(initial_state, goal_state, blocks);

    cout << "\n========= FINAL PLAN =========\n";
    int step = 1;
    for (auto& a : plan)
        cout << step++ << ". " << a.str() << endl;

    return 0;
}
